lib/ansible/plugins/lookup/avi.py

```diff
@@ -0,0 +1,89 @@
+# python 3 headers, required if submitting to Ansible
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+lookup: avi
+author: Eric Anderson <eanderson@avinetworks.com>
+version_added: 2.8
+short_description: read file contents
+description:
+    - this lookup plugin will get data from the controller object you request
+options:
+    obj_type:
+        description:
+            - type of object to query
+    obj_name:
+        description:
+            - name of the object to query
+    obj_uuid:
+        description:
+            - UUID of the object to query
+extends_documentation_fragment: avi
+"""
+
+EXAMPLES = """
+- debug: msg="{{ lookup('avi', obj_name='vs1', obj_type='virtualservice',
+    tenant='admin', controller_ip='controller_ip', username='username', password='password') }}"
+- debug: msg="{{ lookup('avi', obj_uuid='virtualservice-5c0e183a-690a-45d8-8d6f-88c30a52550d',
+    obj_type='virtualservice', tenant='admin', controller_ip='controller_ip', username='username',
+    password='password') }}"
+"""
+
+RETURN = """
+_raw:
+    description:
+        - object data structure from Avi controller as a dict
+"""
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.plugins.lookup import LookupBase
+from ansible.utils.display import Display
+from ansible.module_utils.networks.avi.avi_api import ApiSession
+
+display = Display()
+
+
+class LookupModule(LookupBase):
+    def run(self, terms, variables=None, **kwargs):
+        try:
+            session_params = {
+                'controller_ip': kwargs.get('controller_ip', None),
+                'username': kwargs.get('username', None),
+                'password': kwargs.get('password', None),
+                'token': kwargs.get('token', None),
+                'tenant': kwargs.get('tenant', None),
+                'tenant_uuid': kwargs.get('tenant_uuid', None),
+                'verify': kwargs.get('verify', None),
+                'port': kwargs.get('port', None),
```

Comment:
I would suggest we don't expose everything. let us retrict it to the options allowed in the ansible modules.

Reply:
The problem was is how they are submitted via lookup, there's no way to seperate them unless we unlike every other lookup plugin do a dict only var for auth. Our session function does not support kwargs

Reply:
Another option is to rewrite the module_utils.avi to include ability to process requests from plugins and not from modules...but that's also why Google uses a MockupModule class to emulate a module value for sessions.

```class GcpMockModule(object):
    def __init__(self, params):
        self.params = params

    def fail_json(self, *args, **kwargs):
        raise AnsibleError(kwargs['msg'])

    def raise_for_status(self, response):
        try:
            response.raise_for_status()
        except getattr(requests.exceptions, 'RequestException'):
            self.fail_json(msg="GCP returned error: %s" % response.json())


class GcpFileLookup():
    def get_file_contents(self, module):
        auth = GcpSession(module, 'storage')```